Google.PowerShell/Storage/GcsObject.cs

```diff
@@ -243,27 +276,99 @@ protected override void ProcessRecord()
+        /// then the bucket should have A\C.txt and A\B\D.txt
+        /// </summary>
+        private void UploadDirectory(string directory, Dictionary<string, string> metadataDict, string gcsObjectNamePrefix)
+        {
+            if (string.IsNullOrWhiteSpace(directory) || !Directory.Exists(directory))
+            {
+                return;
+            }
+
+            // We have to append / at the end to create a directory on the cloud.
+            string gcsObjectNamePrefixWithSlash = gcsObjectNamePrefix + "/";
+
+            if (TestObjectExists(Service, Bucket, gcsObjectNamePrefixWithSlash) && !Force.IsPresent)
+            {
+                throw new PSArgumentException(
+                    $"Storage object '{gcsObjectNamePrefixWithSlash}' already exists. Use -Force to overwrite.");
+            }
+
+            // Create a directory on the cloud.
+            string objContentType = GetContentType(null, metadataDict, UTF8TextMimeType);
+            byte[] contentBuffer = Encoding.UTF8.GetBytes(Contents);
+            Stream contentStream = new MemoryStream(contentBuffer);
+            UploadStreamToGcsObject(contentStream, objContentType, metadataDict, gcsObjectNamePrefixWithSlash);
+
+            foreach (string file in Directory.EnumerateFiles(directory))
+            {
+                // Upload each individual file.
+                string fileName = Path.GetFileName(file);
+                string fileWithGcsObjectNamePrefix = Path.Combine(gcsObjectNamePrefix, fileName);
+                // We have to replace \ with / so it will be created with correct folder structure.
+                fileWithGcsObjectNamePrefix = ConvertLocalToGcsFolderPath(fileWithGcsObjectNamePrefix);
+                UploadStreamToGcsObject(new FileStream(file, FileMode.Open),
+                    GetContentType(ContentType, metadataDict, InferContentType(file)),
+                    metadataDict,
+                    ConvertLocalToGcsFolderPath(fileWithGcsObjectNamePrefix));
+            }
+
+            foreach (string subDirectory in Directory.EnumerateDirectories(directory))
+            {
+                // Recursively upload subfolder.
+                string subDirectoryName = Path.GetFileName(subDirectory);
+                string subDirectoryWithGcsObjectNamePrefix = Path.Combine(gcsObjectNamePrefix, subDirectoryName);
+                UploadDirectory(subDirectory,
+                    metadataDict,
+                    ConvertLocalToGcsFolderPath(subDirectoryWithGcsObjectNamePrefix));
+            }
+        }
+
+        /// <summary>
+        /// Upload a GCS object using a stream.
+        /// </summary>
+        private void UploadStreamToGcsObject(Stream contentStream, string objContentType, Dictionary<string,string> metadataDict, string objectName)
+        {
+            if (contentStream == null)
+            {
+                return;
```

Comment:
While we may be safe in that all callpaths to this function give us a valid stream, maybe we should allow this case. e.g. just create a zero-length MemoryStream so that it will create a 0-byte file on GCS? It seems strange that the function would silently do nothing, as opposed to throwing an InvalidArgumentException.